packages/cli/generation/remote-generation/remote-workspace-runner/src/SourceUploader.ts

```diff
@@ -0,0 +1,141 @@
+import { AbsoluteFilePath, join, RelativeFilePath } from "@fern-api/fs-utils";
+import { ApiDefinitionSource, SourceConfig } from "@fern-api/ir-sdk";
+import { loggingExeca } from "@fern-api/logging-execa";
+import { InteractiveTaskContext } from "@fern-api/task-context";
+import { FernWorkspace, IdentifiableSource } from "@fern-api/workspace-loader";
+import { FernRegistry as FdrAPI } from "@fern-fern/fdr-cjs-sdk";
+import { readFile, unlink } from "fs/promises";
+import tmp from "tmp-promise";
+
+const PROTOBUF_ZIP_FILENAME = "proto.zip";
+
+export type SourceType = "asyncapi" | "openapi" | "protobuf";
+
+export class SourceUploader {
+    private context: InteractiveTaskContext;
+    private sources: Record<string, IdentifiableSource>;
+    private sourceTypes: Set<SourceType>;
+
+    constructor(context: InteractiveTaskContext, workspace: FernWorkspace) {
+        this.context = context;
+        this.sources = Object.fromEntries(workspace.getSources().map((source) => [source.id, source]));
+        this.sourceTypes = new Set<SourceType>(Object.values(this.sources).map((source) => source.type));
+    }
+
+    public getSourceTypes(): Set<SourceType> {
+        return this.sourceTypes;
+    }
+
+    public getFdrApiDefinitionSources():
```

Comment:
can we make only the primary method that is used externally public, otherwise private?

Reply:
That's already the case / this is used publicly - it maps the sources into the format that FDR expects in the register call:

```typescript
const sourceUploader = new SourceUploader(interactiveTaskContext, workspace);
const sources = sourceUploader.getFdrApiDefinitionSources();

const apiDefinition = convertIrToFdrApi({ ir, snippetsConfig: {} });
const response = await fdr.api.v1.register.registerApiDefinition({
    orgId: organization,
    apiId: ir.apiName.originalName,
    definition: apiDefinition,
    sources
});
```